### ISSUE TITLE
Object.assign allows unsafe assignment

### ISSUE THREAD

**Issue by d07RiV** [microsoft/TypeScript#59477]
### 🔎 Search Terms

Object.assign

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about Object.assign

### ⏯ Playground Link

https://www.typescriptlang.org/play/?#code/MYewdgzgLgBAZiEMC8MDeMAeAuGByACwFMAbEkPGAXwCgaB5AIwCsjgoA6AQwggEsA5mAAUCEABp0NGFlwBmGlQCUdMR0wcoIADIgA7kQBOAYR5FhSoA

### 💻 Code

```ts
const foo = { x: 'hello' }

Object.assign(foo, {
  x: 3
})

foo.x.toLowerCase()
```

### 🙁 Actual behavior

Typescript allows assignment of incompatible types to `foo`

### 🙂 Expected behavior

It should report an error when trying to assign a number to string

### Additional information about the issue

Shouldn't `Object.assign` be typed as `<T>assign(dst: T, ...src: Partial<T>[])`?

I understand that it would prevent its usage with an empty object as first argument, but shouldn't that be done using spread operator instead? Not performing any type checking in the use case where `Object.assign` has no alternatives seems like a big concern.

Ideally it would skip type checking if the first argument is a literal, but as I understand that's not possible currently.

**Comment by RyanCavanaugh** [comment]
Unfortunately there are also idiomatic applications of `Object.assign` where the mutation to the first argument isn't relevant (and in fact might incur type violations, as shown here), e.g.:
```ts
type Options = {
  x?: number | string,
  y?: number
}

// Set up default options
const result = Object.assign({ x: 3, y: 2 }, someOptions);
```
Since `assign` is polyfillable but spread syntax isn't, there are still a lot of code bases using it where spread would be more appropriate

**Comment by d07RiV** [comment]
I see, I didn't consider polyfill compatibility.

I don't suppose there's any way to accomplish this?

> Ideally it would skip type checking if the first argument is a literal, but as I understand that's not possible currently.

**Comment by RyanCavanaugh** [comment]
> I don't suppose there's any trick that would differentiate between a literal and a variable argument?

Nope
